### ISSUE TITLE
policy files: make `package` optional

### ISSUE THREAD

**Issue by srenatus** [open-policy-agent/opa#5248]
With #4660, we could allow omitting the `package` statement of a policy.

```rego
package authz

allow := true
```
⏬ 
```rego
authz.allow := true
```

It's to be decided if we want to allow non-ref rules. As it stands, the shortest-possible rule still is `data.a.b` (since it needs to live in a package, `a`) Without requiring a package, the shorted-possible rule could become `data.b`.

This will come it with its own feature cap.

Obviously, we'll need to check code locations where having a package is assumed.

**Comment by anderseknert** [comment]
Is there a better motivation available for this than "we could"? 😅 

Maybe it's just old habit, but I rather like having a recognizable structure for a Rego policy.

**Comment by srenatus** [comment]
Yeah it's hard to shake the habit. But I vaguely remember having come to OPA as a novice user once, and it took a while for me to understand what `package a.b.foo` means and why it matters. So with this change, it would be come an optional thing, basically:

> A Rego file contains rule definitions.
> To collect those rules under a common "authz" prefix, use `package authz`.

**Comment by philipaconrad** [comment]
@srenatus I rather like where that takes the language. :smile: The whole point of a package declaration then becomes just abbreviating the rule names.

**Comment by anderseknert** [comment]
The problem is that for any option where there are two or more ways of doing something basic or fundamental, anyone doing introductions to Rego, demos, etc.. will always have to answer "why did you use X over there but not here?"... which almost always distracts from the topic at hand.

We've seen that in the past with e.g. `default allow = true` — "why did you use `:=` everywhere else but `=` here?" and have tried to fix that, in that case by allowing `:=` for default constructs, and "softly" deprecating the alternative. 

This feels like a minor convenience to power users, with the risk of adding a burden for people teaching/learning Rego. But again, could be that the ref-heads change will have that effect anyway, and this will make it nicer to work with that. See "old habits" mentioned above :)

**Comment by srenatus** [comment]
So just recently someone who came new to the language assumed that `package terraform.analysis` was an import, enabling them to use some special TF features. I guess if `package` wasn't a thing at all, that couldn't have happened.

OTOH, I wonder if more people thought that if `package` became optional... 🙃
